mantidproject/mantid
# Change SaveIntegratedWorkspace parameter on SANSTubeCalibration algorithm to take a path to a directory

### Description of work

#### Summary of work

Addresses third item on issue #35981. The algorithm now takes the path to a directory that will be used as the location for saving (and attempting to load) the integrated input workspaces. I noticed that it's still possible to enter the path to a file when using the `FilePath` property with `FileAction.OptionalDirectory`, so, to avoid confusion, I have included validation that will throw an error if the filepath entered is not a directory.



Refs #35981. 




### To test:


This is a long-running algorithm, so the following tests are intended to result in an error that will cut the algorithm run short but allow enough of it to run to check the loading:

1) Search for the `SANSTubeCalibration` algorithm from the Algorithm pane and click Execute.
1) In the dialog box, set the following parameters:
    - `StripPositions`: 920
    - `DataFiles`: SANS2D00069117.nxs
    - `Rear`: unticked
    - `Threshold`: 1000
    - `SkipTubesOnError`: unticked
    - `SaveIntegratedWorkspaces`: set this to a directory to save the integrated input workspace to
1) Run the calibration.
1) In the messages pane you should eventually see a message `Loading and integrating data from SANS2D00069117.nxs`.
1) After the load has completed and the calibration has begun the run will fail with an error similar to _"Error in execution of algorithm SANSTubeCalibration: Failure attempting to calibrate tube 0 - Too few usable points in tube to perform fitting."_. However, at this point the integrated input workspace should have been saved to the directory you set for `SaveIntegratedWorkspaces`. It should have been saved as `tubeCalibSaved_SANS2D00069117.nxs`.
1) Now test that this file is picked up for loading next time the algorithm is run. To do this, clear your workspace list and then re-run the algorithm with the same parameters as before.
1) The calibration run will again fail with the same error as in step 5, however from the load step you should see this message in the messages pane: `Loaded saved file from [path_to_specified_save_directory]\tubeCalibSaved_SANS2D00069117.nxs`.
1) Now check that the file will be picked up for loading from one of your Mantid paths if it isn't in the directory you specified. To do this, move file `tubeCalibSaved_SANS2D00069117.nxs` out of it's current location and to one of your other Mantid paths/your Mantid default save directory.
1) Clear your workspace list and re-run the algorithm with the same parameters as before. This time in the messages pane you should see: `Loaded saved file from tubeCalibSaved_SANS2D00069117.nxs`.
1) Check the validation of the SaveIntegratedWorkspace parameter by trying to enter some invalid filepaths and paths to locations that are not directories.





---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
